src/Poseidon/CLI/Jannocoalesce.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Poseidon.CLI.Jannocoalesce where
+
+import           Poseidon.Janno         (JannoRow (..), JannoRows (..),
+                                         readJannoFile, writeJannoFile)
+import           Poseidon.Package       (PackageReadOptions (..),
+                                         defaultPackageReadOptions,
+                                         getJointJanno,
+                                         readPoseidonPackageCollection)
+import           Poseidon.Utils         (PoseidonException (..), PoseidonIO,
+                                         logInfo)
+
+import           Control.Monad          (forM)
+import           Control.Monad.Catch    (MonadThrow, throwM)
+import           Control.Monad.IO.Class (liftIO)
+import qualified Data.ByteString.Char8  as BSC
+import qualified Data.Csv               as Csv
+import qualified Data.HashMap.Strict    as HM
+import           System.Directory       (createDirectoryIfMissing)
+import           System.FilePath        (takeDirectory)
+
+-- the source can be a single janno file, or a set of base directories as usual.
+data JannoSourceSpec = JannoSourceSingle FilePath | JannoSourceBaseDirs [FilePath]
+
+data JannoCoalesceOptions = JannoCoalesceOptions
+    { _jannocoalesceSource           :: JannoSourceSpec
+    , _jannocoalesceTarget           :: FilePath
+    , _jannocoalesceOutSpec          :: Maybe FilePath -- Nothing means "in place"
+    , _jannocoalesceFillColumns      :: [String] -- empty list means All
+    , _jannocoalesceOverwriteColumns :: Bool
+    }
+
+runJannocoalesce :: JannoCoalesceOptions -> PoseidonIO ()
+runJannocoalesce (JannoCoalesceOptions sourceSpec target outSpec fields overwrite) = do
+    JannoRows sourceRows <- case sourceSpec of
+        JannoSourceSingle sourceFile -> readJannoFile sourceFile
+        JannoSourceBaseDirs sourceDirs -> do
+            let pacReadOpts = defaultPackageReadOptions {
+                      _readOptIgnoreChecksums      = True
+                    , _readOptGenoCheck            = False
+                    , _readOptIgnoreGeno           = True
+                    , _readOptOnlyLatest           = True
+                }
+            getJointJanno <$> readPoseidonPackageCollection pacReadOpts sourceDirs
+    JannoRows targetRows <- readJannoFile target
+    newJanno <- forM targetRows $ \targetRow -> do
+        let posId = jPoseidonID targetRow
+            sourceRowCandidates = filter (\r -> jPoseidonID r == posId) sourceRows
+        case sourceRowCandidates of
+            [] -> return targetRow
+            [keyRow] -> mergeRow targetRow keyRow fields overwrite
+            _ -> throwM $ PoseidonGenericException $ "source file contains multiple rows with key " ++ posId
+    let outPath = maybe target id outSpec
```

Comment:
I'm not sure, but I think overwriting in place can fail because of laziness. Didn't test it, but I feel this could be an issue. Maybe there should be some forced strictness via bang patterns in line 45? Or is this not an issue here for some reason?

Reply:
Hmm, that is a good question. I do not know, actually. I think this comes down to whether Cassava loads files lazily or not. I would hope that it doesn't. Lazy-IO is considered bad practice in Haskell, usually. I don't really know how to test this in a unit test. In a Golden test it's probably possible, but those are so difficult to write. I would suggest you perhaps try it out by hand, and we just trust that it works then?

Reply:
OK, I checked manually and it seems to work just fine.